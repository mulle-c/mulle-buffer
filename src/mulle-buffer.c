//
//  mulle_buffer.c
//  mulle-buffer
//
//  Created by Nat! on 04/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#include "mulle-buffer.h"

#include "include-private.h"

#include <ctype.h>


struct mulle_buffer   *mulle_buffer_create( struct mulle_allocator *allocator)
{
   struct mulle_buffer  *buffer;

   if( ! allocator)
      allocator = &mulle_default_allocator;

   buffer = mulle_allocator_malloc( allocator, sizeof( struct mulle_buffer));
   mulle_buffer_init( buffer, allocator);
   return( buffer);
}


static inline unsigned int   hex( unsigned int c)
{
   assert( c >= 0 && c <= 0xf);
   return( c >= 0xa ? c + 'a' - 0xa : c + '0');
}


void  mulle_buffer_hexdump_line( struct mulle_buffer *buffer,
                                 uint8_t *bytes,
                                 unsigned int n,
                                 size_t counter,
                                 unsigned int options)
{
   uint8_t        *memo;
   uint8_t        *s;
   unsigned int   i;
   unsigned int   value;

   memo = bytes;

   s = mulle_buffer_advance( buffer, 10);

   if( ! (options & mulle_buffer_hexdump_offset))
   {
      *s++ = (uint8_t) hex( counter >> 28 & 0xF);
      *s++ = (uint8_t) hex( counter >> 24 & 0xF);
      *s++ = (uint8_t) hex( counter >> 20 & 0xF);
      *s++ = (uint8_t) hex( counter >> 16 & 0xF);

      *s++ = (uint8_t) hex( counter >> 12 & 0xF);
      *s++ = (uint8_t) hex( counter >> 8 & 0xF);
      *s++ = (uint8_t) hex( counter >> 4 & 0xF);
      *s++ = (uint8_t) hex( counter >> 0 & 0xF);

      *s++ = ' ';
      *s++ = ' ';
   }

   if( ! (options & mulle_buffer_hexdump_hex))
   {
      s = mulle_buffer_advance( buffer, 3 * 8);
      for( i = 0; i < 8; i++)
      {
         if( i < n)
         {
            value = *bytes++;
            *s++ = (uint8_t) hex( value >> 4);
            *s++ = (uint8_t) hex( value & 0xF);
         }
         else
         {
            *s++ = ' ';
            *s++ = ' ';
         }
         *s++ = ' ';
      }

      mulle_buffer_add_byte( buffer, ' ');

      s = mulle_buffer_advance( buffer, 3 * 8);
      for( ; i < 16; i++)
      {
         if( i < n)
         {
            value = *bytes++;

            *s++ = (uint8_t) hex( value >> 4);
            *s++ = (uint8_t) hex( value & 0xF);
         }
         else
         {
            *s++ = ' ';
            *s++ = ' ';
         }
         *s++ = ' ';
      }
   }

   if( ! (options & mulle_buffer_hexdump_ascii))
   {
      mulle_buffer_add_byte( buffer, ' ');
      mulle_buffer_add_byte( buffer, '|');

      bytes = memo;

      for( i = 0; i < 16; i++)
      {
         if( i < n)
         {
            value = *bytes++;
            if( ! isprint( value))
               value = '.';
            mulle_buffer_add_byte( buffer, (uint8_t) value);
         }
         else
            mulle_buffer_add_byte( buffer, ' ');
      }

      mulle_buffer_add_byte( buffer, '|');
   }
}


void  mulle_buffer_hexdump( struct mulle_buffer *buffer,
                            uint8_t *bytes,
                            size_t length,
                            size_t counter,
                            unsigned int options)
{
   size_t   lines;
   size_t   full_lines;
   size_t   remainder;
   size_t   i;

   lines      = (length + 15) / 16;
   full_lines = length / 16;

   for( i = 0; i < full_lines; i++)
   {
      mulle_buffer_hexdump_line( buffer, bytes, 16, counter, options);
      mulle_buffer_add_byte( buffer, '\n');
      counter += 16;
      bytes   += 16;
   }

   if( i < lines)
   {
      remainder = length - full_lines * 16;
      mulle_buffer_hexdump_line( buffer, bytes, (unsigned int) remainder, counter, options);
      mulle_buffer_add_byte( buffer, '\n');
   }
}


int   mulle_flushablebuffer_done( struct mulle_flushablebuffer *buffer)
{
   int   rval;

   rval = _mulle_flushablebuffer_flush( (struct _mulle_flushablebuffer *) buffer);
   if( rval)
      return( rval);

   _mulle_buffer_done( (struct _mulle_buffer *) buffer,
                      mulle_buffer_get_allocator( (struct mulle_buffer *) buffer));
   return( 0);
}


int   mulle_flushablebuffer_destroy( struct mulle_flushablebuffer *buffer)
{
   int   rval;

   rval = _mulle_flushablebuffer_flush( (struct _mulle_flushablebuffer *) buffer);
   if( rval)
      return( rval);

   _mulle_buffer_destroy( (struct _mulle_buffer *) buffer,
                      mulle_buffer_get_allocator( (struct mulle_buffer *) buffer));
   return( 0);
}

