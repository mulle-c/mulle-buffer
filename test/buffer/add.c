#include <mulle-buffer/mulle-buffer.h>
#include <mulle-testallocator/mulle-testallocator.h>


#include <stdio.h>



static void   simple_fill_test()
{
   struct mulle_buffer   *buffer;
   unsigned int           i;
   size_t                 len;

   buffer = mulle_buffer_create( NULL);

   for( i = 0; i < 100000; i++)
      mulle_buffer_add_byte( buffer, 'a' + i % 26);
   printf( "simple_fill_test: %ld\n", mulle_buffer_get_length( buffer));

   mulle_buffer_memset( buffer, 'z', 1000);
   mulle_buffer_zero_last_byte( buffer);

   len = strlen( mulle_buffer_get_bytes( buffer));
   printf( "simple_fill_test: %ld\n", len);

//   printf( "simple_fill_test: %lx\n", mulle_hash( mulle_buffer_get_bytes( buffer),
//                                                  len));

   mulle_buffer_add_bytes( buffer, "hello", 5);

   printf( "simple_fill_test: %ld\n", mulle_buffer_get_length( buffer));

   printf( "simple_fill_test: %ld\n", mulle_buffer_get_staticlength( buffer));

   mulle_buffer_reset( buffer);
   printf( "simple_fill_test: %ld\n", mulle_buffer_get_length( buffer));

   mulle_buffer_destroy( buffer);
}


static void   simple_static_test()
{
   struct mulle_buffer   buffer;

   // this can't grow
   mulle_buffer_init_inflexible_with_static_bytes( &buffer,
                                                   "hello world",
                                                   12);
   // skip over first 11 bytes
   mulle_buffer_advance( &buffer, 12);

   // test that we can't write to it
   mulle_buffer_add_byte( &buffer, 'x');
   mulle_buffer_memset( &buffer, 'y', 3);

   printf( "simple_static_test: %ld\n", mulle_buffer_get_length( &buffer));
   printf( "simple_static_test: %s\n", (char *) mulle_buffer_get_bytes( &buffer));
   printf( "simple_static_test: %ld\n", mulle_buffer_get_staticlength( &buffer));

   mulle_buffer_reset( &buffer);
   printf( "simple_fill_test: %ld\n", mulle_buffer_get_length( &buffer));

   mulle_buffer_done( &buffer);
}

static void   if_empty_test( void)
{
   mulle_buffer_do( buffer)
   {
      mulle_buffer_add_string_if_not_empty( buffer, "A");
      mulle_buffer_add_string_if_empty( buffer, "B");
      mulle_buffer_add_string_if_not_empty( buffer, "C");
      mulle_buffer_add_string_if_empty( buffer, "D");
      mulle_buffer_make_string( buffer);
      printf( "if_empty_test: %s\n", (char *) mulle_buffer_get_bytes( buffer));
   }
}



int  main()
{
   simple_fill_test();
   mulle_testallocator_reset();
   simple_static_test();
   if_empty_test();

   return( 0);
}

