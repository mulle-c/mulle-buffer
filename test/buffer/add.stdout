simple_fill_test: 100000
simple_fill_test: 101000
simple_fill_test: 101005
simple_fill_test: 0
simple_fill_test: 0
simple_static_test: 12
simple_static_test: hello world
simple_static_test: 12
simple_fill_test: 0
if_empty_test: BC
